src/peer_connection.cpp

```diff
@@ -1973,7 +1974,17 @@ namespace libtorrent {
 			// instead.
 			if (super_seeded_piece(index))
 			{
-				superseed_piece(index, t->get_piece_to_super_seed(m_have_piece));
+#ifndef TORRENT_DISABLE_LOGGING
+				if (should_log(peer_log_alert::incoming))
+					peer_log(peer_log_alert::incoming, "HAVE"
+						, "got HAVE message for superseeded piece %d"
+						, static_cast<int>(index));
+#endif
+				// Don't replace the piece if we're above the current target
+				if (target_superseed_pieces() < m_superseed_pieces.size())
+					m_superseed_pieces.erase(std::remove(m_superseed_pieces.begin(), m_superseed_pieces.end(), index)
```

Comment:
`std::remove()` was introduced in C++17, and this branch of libtorrent is locked in to C++11. `RC_2_0` is C++14 and `master` is C++17.

However, it seems the `m_superseed_pieces` my be better suited as a `set`. As far as I can tell, we only use it to insert piece indices, and look up whether an index is in there, and then remove an index. (it could also be a bitmask, but that would most likely be less efficient)

Reply:
I could probabaly rework it as a `set` instead, though might need to change how we handle peers that filter sending redundant `HAVE` messages.

Previously we dealt with this by waiting for a peer to be uninterested for 10 seconds, and then offering new pieces until the peer becomes interested again. I adapted this for use with multiple pieces by using a `vector` - adding the newly offered piece to the beginning, and resizing off the oldest piece/s at the end when it grows too big. 

Using a `set`, I guess we would then clear it whenever we try offering a new piece to such a peer, as this would be closest to the current behavior. Such a peer wouldn't be able to benefit from having multiple pieces to choose from anyway. Thankfully, so far in testing, I've noticed very few peers that seem to behave this way anwyay.

There also might be some weirdness when a peer only has a few pieces to go. If I'm reading it correctly, `torrent::get_piece_to_super_seed()` can still return a piece that is already being superseeded to the peer if there are no other pieces left to pick from. We would have to take this into account when calculating a target number of pieces to superseed, otherwise we would keep adding duplicate pieces, but the set size would never grow to the target as there aren't enough unique needed pieces left. This is probabaly something that needs addressing anyway. As it stands, while functional, the `vector` would be filling with duplicate pieces as the peer nears completion and we would be re-sending out `HAVE` messages for the last few pieces that we've already made available.